Sources/UIAlertControllerView.swift

```diff
@@ -43,48 +57,103 @@ class UIAlertControllerView: UIView {
             return button
         }
 
+        clipsToBounds = true
         backgroundColor = .white
         layer.cornerRadius = 3
 
-        header.sizeToFit()
-        header.frame.height += verticalPadding
-        header.layer.contentsGravity = .top
         addSubview(header)
+        text.map { addSubview($0) }
+        buttons.forEach { addSubview($0) }
+
+        subviewWidth = {
+            let minWidth: CGFloat = 300
+            let maxWidth: CGFloat
+            if UIScreen.main.isPortrait {
+                maxWidth = UIScreen.main.bounds.width - 4 * horizontalPadding
+            } else {
+                maxWidth = max(minWidth, UIScreen.main.bounds.width / 1.5)
+            }
 
-        buttons.forEach{ addSubview($0) }
+            subviews.forEach { $0.sizeToFit() }
+            guard let largestSubviewWidth = subviews.map({ $0.frame.width }).max() else {
+                assertionFailure("no subviews exist to determine largestSubviewWidth")
+                return minWidth
+            }
+            return min(maxWidth, max(largestSubviewWidth, minWidth))
+        }()
     }
 
     override func layoutSubviews() {
-        switch style {
-        case .actionSheet: layoutAsActionSheet()
-        default: assertionFailure(
-            "The UIAlertControllerStyle style, \(style), is not implemented yet!")
+        subviews.forEach { $0.sizeToFit() }
+
+        header.frame.origin = CGPoint(x: horizontalPadding, y: verticalPadding)
+        header.bounds.width = subviewWidth
+        header.sizeToFit()
+        header.textAlignment = .left // sizeToFit() seems to change textAlignment
+
+        text?.frame.origin.x = horizontalPadding
+        text?.frame.width = subviewWidth
+        text?.sizeToFit()
+        text?.frame.minY = header.frame.maxY + verticalPadding
+
+        if style == .alert {
+            let totalButtonsWidth = buttons.reduce(0, { $0 + $1.frame.width }) + horizontalGap * CGFloat(buttons.count - 1)
+            if totalButtonsWidth < subviewWidth {
+                return layoutButtonsHorizontally()
+            }
         }
+        return layoutButtonsVertically()
+
     }
 
-    private func layoutAsActionSheet() {
-        header.frame.origin = CGPoint(x: horizontalPadding, y: verticalPadding)
+    private func layoutButtonsVertically() {
+        buttons.enumerated().forEach({ (arg) in
+            let (index, button) = arg
+
+            let fullButtonWidth = subviewWidth + horizontalPadding * 2
+
+            switch style {
+            case .actionSheet:
+                button.titleLabelOriginX = horizontalPadding
+            case .alert:
+                button.titleLabelOriginX = fullButtonWidth - horizontalPadding - button.bounds.width
+            }
+
+            button.frame.size.width = fullButtonWidth
+            button.frame.size.height += verticalPadding
+
+            if button == buttons.first {
```

Comment:
nit:
We also have the `index` so this could also be:
```suggestion
            if index == 0 {
```

However, whichever you prefer. I find this equally easy to read. Also in the function below.